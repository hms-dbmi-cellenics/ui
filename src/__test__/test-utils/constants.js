export default {
  EXPERIMENT_ID: 'testae48e318dab9a1bd0bexperiment',
  CLONED_EXPERIMENT_ID: 'testae48e318dab9a1bd0bexperclone',
  PROJECT_WITH_SAMPLES_UUID: '380c53b4-mock-test-8091-4c82336d6d49',
  PROJECT_WITHOUT_SAMPLE_UUID: '4761594b-mock-test-ba89-c2c812d39bb5',
  EXPERIMENT_NAME: 'Test Experiment',
  PROJECT_ID: 'test24b6-8600-test-mock-63822d2fmock',
  SAMPLE_ID: 'test9188-d682-test-mock-cb6d644cmock',
  API_ENDPOINT: /^http?:\/\/localhost:3000.*$/,
  S3_ENDPOINT: 'http://mock.s3.amazonaws.com',
  // pragma: allowlist secret
  JWT_TOKEN: 'Bearer eyJhbGciOiJIUzI1NiIsInR5cCI6Ikpfake.eyJzdWIiOiIxMjM0NTY3ODkwIiwifakeZSI6IkpvaG4gRG9lIiwiaWF0IjoxNTfakeM5MDIyfQ.fakexwRfakeKKF2QT4fwpMeJf36POk6yJV_adQsfake',
  MOCK_DATETIME: '0000-00-00T00:00:00.000Z',
  ETAG: '380c53b4-mock-etag-8091-4c82336d6d49',
};
