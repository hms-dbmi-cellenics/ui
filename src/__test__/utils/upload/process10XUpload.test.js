import _ from 'lodash';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import waitForActions from 'redux-mock-store-await-actions';
import axios from 'axios';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';
import { waitFor } from '@testing-library/react';

import {
  SAMPLES_CREATED, SAMPLES_FILE_UPDATE, SAMPLES_SAVED, SAMPLES_SAVING, SAMPLES_VALIDATING_UPDATED,
} from 'redux/actionTypes/samples';
import initialSampleState, { sampleTemplate } from 'redux/reducers/samples/initialState';
import initialExperimentState, { experimentTemplate } from 'redux/reducers/experiments/initialState';

import UploadStatus from 'utils/upload/UploadStatus';

import processUpload from 'utils/upload/processUpload';

import validate from 'utils/upload/validate10x';
import pushNotificationMessage from 'utils/pushNotificationMessage';
import { sampleTech } from 'utils/constants';
import mockFile from '__test__/test-utils/mockFile';
import loadAndCompressIfNecessary from 'utils/upload/loadAndCompressIfNecessary';

enableFetchMocks();

const getValidFiles = (cellrangerVersion, compressed = true) => {
  const featuresFilename = cellrangerVersion === 'v2' ? 'genes.tsv.gz' : 'features.tsv.gz';

  let fileList = [
    {
      name: `${featuresFilename}`,
      fileObject: mockFile(featuresFilename, 'WT13'),
      upload: { status: UploadStatus.UPLOADING },
      errors: '',
      compressed,
      valid: true,
    },
    {
      name: 'barcodes.tsv.gz',
      fileObject: mockFile('barcodes.tsv.gz', 'WT13'),
      upload: { status: UploadStatus.UPLOADING },
      errors: '',
      compressed,
      valid: true,
    },
    {
      name: 'matrix.mtx.gz',
      fileObject: mockFile('matrix.mtx.gz', 'WT13'),
      upload: { status: UploadStatus.UPLOADING },
      errors: '',
      compressed,
      valid: true,
    },
  ];

  fileList = fileList.map((file) => ({ ...file, size: file.fileObject.size }));
  return fileList;
};

const sampleType = sampleTech['10X'];
const mockSampleUuid = 'sample-uuid';
const mockExperimentId = 'project-uuid';
const sampleName = 'mockSampleName';

const mockUnrelatedSampleUuid = 'unrelated-sample-uuid';
const mockUnrelatedExperimentId = 'unrelated-experiment-id';

const initialState = {
  experiments: {
    ...initialExperimentState,
    [mockExperimentId]: {
      ...experimentTemplate,
      id: mockExperimentId,
      sampleIds: [mockSampleUuid],
    },
  },
  samples: {
    ...initialSampleState,
    ids: [mockSampleUuid],
    meta: {
      loading: true,
      error: false,
    },
    [mockSampleUuid]: {
      ...sampleTemplate,
      uuid: mockSampleUuid,
      name: sampleName,
      experimentId: mockExperimentId,
    },
    [mockUnrelatedSampleUuid]: {
      ...sampleTemplate,
      uuid: mockUnrelatedSampleUuid,
      name: sampleName,
      experimentId: mockUnrelatedExperimentId,
    },
  },
};

const mockStore = configureMockStore([thunk]);

jest.mock('utils/upload/loadAndCompressIfNecessary',
  () => jest.fn().mockImplementation(
    (file) => {
      if (!file.valid) {
        return Promise.reject(new Error('error'));
      }
      return Promise.resolve('loadedGzippedFile');
    },
  ));

const sampleFileId = 'mockSampleFileId';

jest.mock('uuid', () => ({
  v4: jest.fn(() => sampleFileId),
}));

jest.mock('axios', () => ({
  request: jest.fn(),
}));

jest.mock('utils/pushNotificationMessage');

jest.mock('utils/upload/validate10x');

let store = null;

const mockProcessUploadCalls = () => {
  const sampleId = 'mockSampleId';

  const mockUploadUrlParams = {
    signedUrls: ['theSignedUrl'],
    uploadId: 'some_id',
  };

  fetchMock.mockIf(/.*/, ({ url }) => {
    let result;

    if (url.endsWith(`/v2/experiments/${mockExperimentId}/samples`)) {
      result = { status: 200, body: JSON.stringify({ WT13: sampleId }) };
    }

    // Create sample file
    if (new RegExp(`/v2/experiments/${mockExperimentId}/samples/.*/sampleFiles/.*`).test(url)) {
      result = { status: 200, body: JSON.stringify({}) };
    }

    // Update sample file status
    if (new RegExp(`/v2/experiments/${mockExperimentId}/sampleFiles/.*`).test(url)) {
      result = { status: 200, body: JSON.stringify({}) };
    }

    if (url.endsWith(`/v2/experiments/${mockExperimentId}/sampleFiles/${sampleFileId}/beginUpload`)) {
      result = { status: 200, body: JSON.stringify(mockUploadUrlParams) };
    }

    if (url.endsWith('/v2/completeMultipartUpload')) {
      result = { status: 200, body: JSON.stringify({}) };
    }

    return Promise.resolve(result);
  });
};

describe('processUpload', () => {
  beforeEach(() => {
    jest.clearAllMocks();

    fetchMock.resetMocks();
    fetchMock.doMock();
    mockProcessUploadCalls();

    store = mockStore(initialState);
  });

  it('Uploads and updates redux correctly when there are no errors with cellranger v3', async () => {
    const mockAxiosCalls = [];
    const uploadSuccess = (params) => {
      mockAxiosCalls.push(params);
      return Promise.resolve({ headers: { etag: 'etag-blah' } });
    };

    axios.request.mockImplementationOnce(uploadSuccess)
      .mockImplementationOnce(uploadSuccess)
      .mockImplementationOnce(uploadSuccess);

    const filesList = getValidFiles('v3');

    await processUpload(
      filesList,
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    // Wait for uploads to be made
    await waitForActions(
      store,
      new Array(3).fill({
        type: SAMPLES_FILE_UPDATE,
        payload: { fileDiff: { upload: { status: UploadStatus.UPLOADED } } },
      }),
      { matcher: waitForActions.matchers.containing },
    );

    // Three axios put calls are made
    expect(mockAxiosCalls.length).toBe(3);
    // Each put call is made with the correct information
    expect(mockAxiosCalls[0].data).toBeInstanceOf(Blob);
    expect(mockAxiosCalls[1].data).toBeInstanceOf(Blob);
    expect(mockAxiosCalls[2].data).toBeInstanceOf(Blob);

    const fileUpdateActions = store.getActions().filter(
      (action) => action.type === SAMPLES_FILE_UPDATE,
    );
    const uploadProperties = fileUpdateActions.map((action) => action.payload.fileDiff.upload);
    const uploadingStatusProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADING,
    );
    const uploadedStatusProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADED,
    );

    // Order is respected, SAMPLES_CREATED runs *before* SAMPLES_FILE_UPDATE
    expect(_.map(store.getActions(), 'type')).toEqual([
      ...Array(2).fill(SAMPLES_VALIDATING_UPDATED),
      SAMPLES_SAVING, SAMPLES_CREATED, SAMPLES_SAVED,
      ...Array(6).fill(SAMPLES_FILE_UPDATE),
    ]);

    // There are 3 files actions with status uploading
    expect(uploadingStatusProperties.length).toEqual(3);
    // There are 3 files actions with status uploaded
    expect(uploadedStatusProperties.length).toEqual(3);

    // axios request calls are correct
    expect(axios.request.mock.calls.map((call) => call[0])).toMatchSnapshot();

    // If we trigger axios onUploadProgress it updates the progress correctly
    axios.request.mock.calls[0][0].onUploadProgress({ loaded: filesList[0].fileObject.size / 2 });

    await waitForActions(
      store,
      [{
        type: SAMPLES_FILE_UPDATE,
        payload: { fileDiff: { upload: { status: UploadStatus.UPLOADING, progress: 50 } } },
      }],
      { matcher: waitForActions.matchers.containing },
    );

    expect(fetchMock.mock.calls).toMatchSnapshot('fetch calls');
  });

  it('Uploads and updates redux correctly when there are no errors with cellranger v2', async () => {
    const mockAxiosCalls = [];
    const uploadSuccess = (params) => {
      mockAxiosCalls.push(params);
      return Promise.resolve({ headers: { etag: 'etag-blah' } });
    };

    axios.request.mockImplementationOnce(uploadSuccess)
      .mockImplementationOnce(uploadSuccess)
      .mockImplementationOnce(uploadSuccess);

    const filesList = getValidFiles('v2');

    await processUpload(
      filesList,
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    // Wait for uploads to be made
    await waitForActions(
      store,
      new Array(3).fill({
        type: SAMPLES_FILE_UPDATE,
        payload: { fileDiff: { upload: { status: UploadStatus.UPLOADED } } },
      }),
      { matcher: waitForActions.matchers.containing },
    );

    // Three axios put calls are made
    expect(mockAxiosCalls.length).toBe(3);
    // Each put call is made with the correct information
    expect(mockAxiosCalls[0].data).toBeInstanceOf(Blob);
    expect(mockAxiosCalls[1].data).toBeInstanceOf(Blob);
    expect(mockAxiosCalls[2].data).toBeInstanceOf(Blob);

    const fileUpdateActions = store.getActions().filter(
      (action) => action.type === SAMPLES_FILE_UPDATE,
    );
    const uploadProperties = fileUpdateActions.map((action) => action.payload.fileDiff.upload);
    const uploadingStatusProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADING,
    );
    const uploadedStatusProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADED,
    );

    // There are 3 files actions with status uploading
    expect(uploadingStatusProperties.length).toEqual(3);
    // There are 3 files actions with status uploaded
    expect(uploadedStatusProperties.length).toEqual(3);

    // axios request calls are correct
    expect(axios.request.mock.calls.map((call) => call[0])).toMatchSnapshot();

    // If we trigger axios onUploadProgress it updates the progress correctly
    axios.request.mock.calls[0][0].onUploadProgress({ loaded: filesList[0].fileObject.size / 2 });

    await waitForActions(
      store,
      [{
        type: SAMPLES_FILE_UPDATE,
        payload: { fileDiff: { upload: { status: UploadStatus.UPLOADING, progress: 50 } } },
      }],
      { matcher: waitForActions.matchers.containing },
    );

    expect(fetchMock.mock.calls).toMatchSnapshot('fetch calls');
  });

  it('Updates redux correctly when there are file upload errors', async () => {
    const mockAxiosCalls = [];

    const uploadError = (params) => {
      mockAxiosCalls.push(params);
      return Promise.reject(new Error('Error'));
    };

    axios.request.mockImplementationOnce(uploadError)
      .mockImplementationOnce(uploadError)
      .mockImplementationOnce(uploadError);

    await processUpload(
      getValidFiles('v3'),
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    // Wait for uploads to be made
    await waitForActions(
      store,
      new Array(3).fill({
        type: SAMPLES_FILE_UPDATE,
        payload: { fileDiff: { upload: { status: UploadStatus.UPLOAD_ERROR } } },
      }),
      { matcher: waitForActions.matchers.containing },
    );

    const fileUpdateActions = store.getActions().filter(
      (action) => action.type === SAMPLES_FILE_UPDATE,
    );

    const uploadProperties = fileUpdateActions.map((action) => action.payload.fileDiff.upload);

    const uploadingFileProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADING,
    );

    const errorFileProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOAD_ERROR,
    );

    const uploadedFileProperties = uploadProperties.filter(
      ({ status }) => status === UploadStatus.UPLOADED,
    );

    // There are 3 files actions with status uploading
    expect(uploadingFileProperties.length).toEqual(3);
    // There are 3 files actions with status upload error
    expect(errorFileProperties.length).toEqual(3);
    // There are no file actions with status successfully uploaded
    expect(uploadedFileProperties.length).toEqual(0);

    expect(fetchMock.mock.calls).toMatchSnapshot('fetch calls');
  });

  it('Should not upload files if there are errors creating samples in the api', async () => {
    fetchMock.mockReject(new Error('Error'));

    await processUpload(
      getValidFiles('v3'),
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    // We do not expect uploads to happen
    await waitFor(() => {
      expect(axios.request).not.toHaveBeenCalled();
    });

    expect(fetchMock.mock.calls).toMatchSnapshot('fetch calls');

    // Informs user of error
    expect(pushNotificationMessage).toHaveBeenCalledWith('error', 'We couldn\'t create your sample. Please try uploading it again');
  });

  it('Should not upload files if there are errors beginning the multipart upload in the api', async () => {
    const sampleId = 'mockSampleId';

    fetchMock.mockIf(/.*/, ({ url }) => {
      let result;

      if (url.endsWith(`/v2/experiments/${mockExperimentId}/samples`)) {
        result = { status: 200, body: JSON.stringify({ WT13: sampleId }) };
      }

      if (new RegExp(`/v2/experiments/${mockExperimentId}/samples/.*/sampleFiles/.*`).test(url)) {
        result = { status: 200, body: JSON.stringify({}) };
      }

      if (url.endsWith(`/v2/experiments/${mockExperimentId}/sampleFiles/${sampleFileId}/beginUpload`)) {
        return Promise.reject(new Error('Some error in the api'));
      }

      return Promise.resolve(result);
    });

    await processUpload(
      getValidFiles('v3'),
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    await waitForActions(
      store,
      new Array(3).fill({
        type: SAMPLES_FILE_UPDATE,
        payload: { fileDiff: { upload: { status: UploadStatus.UPLOAD_ERROR } } },
      }),
      { matcher: waitForActions.matchers.containing },
    );

    // Uploads didn't begin
    expect(axios.request).not.toHaveBeenCalled();

    expect(fetchMock.mock.calls).toMatchSnapshot('fetch calls');
  });

  it('Should not upload sample and show notification if uploaded sample is invalid', async () => {
    validate.mockImplementationOnce(
      () => { throw new Error('Some file error'); },
    );

    await processUpload(
      getValidFiles('v2'),
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    // We do not expect uploads to happen
    await waitFor(() => {
      expect(pushNotificationMessage).toHaveBeenCalledTimes(1);
      expect(axios.request).not.toHaveBeenCalled();
    });
  });

  it('Works correctly with files that need to be compressed', async () => {
    const mockAxiosCalls = [];
    const uploadSuccess = (params) => {
      mockAxiosCalls.push(params);
      return Promise.resolve({ headers: { etag: 'etag-blah' } });
    };

    axios.request.mockImplementationOnce(uploadSuccess)
      .mockImplementationOnce(uploadSuccess)
      .mockImplementationOnce(uploadSuccess);

    let finishCompression;
    loadAndCompressIfNecessary.mockReturnValue(new Promise((resolve) => {
      finishCompression = resolve;
    }));

    const filesList = getValidFiles('v3', false);

    processUpload(
      filesList,
      sampleType,
      store.getState().samples,
      mockExperimentId,
      store.dispatch,
    );

    const actionsUpToFileCreation = [
      ...Array(2).fill(SAMPLES_VALIDATING_UPDATED),
      SAMPLES_SAVING, SAMPLES_CREATED, SAMPLES_SAVED,
      ...Array(3).fill(SAMPLES_FILE_UPDATE),
    ];

    // Compression is paused by the test, but the sample files have been created anyways
    // because the order is respected:
    // First: Create sample files
    // After: Compress or begin the actual upload
    await waitFor(() => {
      expect(_.map(store.getActions(), 'type')).toEqual(actionsUpToFileCreation);
    });

    // Upload hasn't begun because compression hasn't finished
    expect(axios.request).not.toHaveBeenCalled();

    // Now we trigger compress
    finishCompression('loadedGzippedFile');

    // After compression, the upload has begun so
    //  new SAMPLES_FILE_UPDATE show up and
    //  axios was called for each file
    await waitFor(() => {
      expect(_.map(store.getActions(), 'type')).toEqual([
        ...actionsUpToFileCreation,
        ...Array(3).fill(SAMPLES_FILE_UPDATE),
      ]);

      expect(mockAxiosCalls.length).toBe(3);
      // Each put call is made with the correct information
      expect(mockAxiosCalls[0].data).toEqual('loadedGzippedFile');
      expect(mockAxiosCalls[1].data).toEqual('loadedGzippedFile');
      expect(mockAxiosCalls[2].data).toEqual('loadedGzippedFile');
    });
  });
});
