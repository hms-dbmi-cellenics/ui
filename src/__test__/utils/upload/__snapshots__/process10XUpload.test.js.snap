// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`processUpload Should not upload files if there are errors beginning the multipart upload in the api: fetch calls 1`] = `
[
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples",
    {
      "body": "[{"name":"WT13","sampleTechnology":"10x","options":{}}]",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/features10x",
    {
      "body": "{"sampleFileId":"mockSampleFileId","size":1024}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/barcodes10x",
    {
      "body": "{"sampleFileId":"mockSampleFileId","size":1024}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/matrix10x",
    {
      "body": "{"sampleFileId":"mockSampleFileId","size":1024}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/sampleFiles/mockSampleFileId/beginUpload",
    {
      "body": "{"size":1024,"metadata":{"cellranger_version":"v3"}}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/sampleFiles/mockSampleFileId/beginUpload",
    {
      "body": "{"size":1024,"metadata":{}}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/sampleFiles/mockSampleFileId/beginUpload",
    {
      "body": "{"size":1024,"metadata":{}}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/features10x",
    {
      "body": "{"uploadStatus":"uploadError"}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "PATCH",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/barcodes10x",
    {
      "body": "{"uploadStatus":"uploadError"}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "PATCH",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/matrix10x",
    {
      "body": "{"uploadStatus":"uploadError"}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "PATCH",
    },
  ],
]
`;

exports[`processUpload Should not upload files if there are errors creating samples in the api: fetch calls 1`] = `
[
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples",
    {
      "body": "[{"name":"WT13","sampleTechnology":"10x","options":{}}]",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
]
`;

exports[`processUpload Updates redux correctly when there are file upload errors: fetch calls 1`] = `
[
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples",
    {
      "body": "[{"name":"WT13","sampleTechnology":"10x","options":{}}]",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/features10x",
    {
      "body": "{"sampleFileId":"mockSampleFileId","size":1024}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/barcodes10x",
    {
      "body": "{"sampleFileId":"mockSampleFileId","size":1024}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/matrix10x",
    {
      "body": "{"sampleFileId":"mockSampleFileId","size":1024}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/sampleFiles/mockSampleFileId/beginUpload",
    {
      "body": "{"size":1024,"metadata":{"cellranger_version":"v3"}}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/sampleFiles/mockSampleFileId/beginUpload",
    {
      "body": "{"size":1024,"metadata":{}}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/sampleFiles/mockSampleFileId/beginUpload",
    {
      "body": "{"size":1024,"metadata":{}}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/features10x",
    {
      "body": "{"uploadStatus":"uploadError"}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "PATCH",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/barcodes10x",
    {
      "body": "{"uploadStatus":"uploadError"}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "PATCH",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/matrix10x",
    {
      "body": "{"uploadStatus":"uploadError"}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "PATCH",
    },
  ],
]
`;

exports[`processUpload Uploads and updates redux correctly when there are no errors with cellranger v2 1`] = `
[
  {
    "data": Blob {},
    "headers": {
      "Content-Type": "application/octet-stream",
    },
    "method": "put",
    "onUploadProgress": [Function],
    "signal": AbortSignal {},
    "url": "theSignedUrl",
  },
  {
    "data": Blob {},
    "headers": {
      "Content-Type": "application/octet-stream",
    },
    "method": "put",
    "onUploadProgress": [Function],
    "signal": AbortSignal {},
    "url": "theSignedUrl",
  },
  {
    "data": Blob {},
    "headers": {
      "Content-Type": "application/octet-stream",
    },
    "method": "put",
    "onUploadProgress": [Function],
    "signal": AbortSignal {},
    "url": "theSignedUrl",
  },
]
`;

exports[`processUpload Uploads and updates redux correctly when there are no errors with cellranger v2: fetch calls 1`] = `
[
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples",
    {
      "body": "[{"name":"WT13","sampleTechnology":"10x","options":{}}]",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/features10x",
    {
      "body": "{"sampleFileId":"mockSampleFileId","size":1024}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/barcodes10x",
    {
      "body": "{"sampleFileId":"mockSampleFileId","size":1024}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/matrix10x",
    {
      "body": "{"sampleFileId":"mockSampleFileId","size":1024}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/sampleFiles/mockSampleFileId/beginUpload",
    {
      "body": "{"size":1024,"metadata":{"cellranger_version":"v2"}}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/sampleFiles/mockSampleFileId/beginUpload",
    {
      "body": "{"size":1024,"metadata":{}}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/sampleFiles/mockSampleFileId/beginUpload",
    {
      "body": "{"size":1024,"metadata":{}}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/completeMultipartUpload",
    {
      "body": "{"parts":[{"ETag":"etag-blah","PartNumber":1}],"uploadId":"some_id","fileId":"mockSampleFileId","type":"sample"}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/completeMultipartUpload",
    {
      "body": "{"parts":[{"ETag":"etag-blah","PartNumber":1}],"uploadId":"some_id","fileId":"mockSampleFileId","type":"sample"}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/completeMultipartUpload",
    {
      "body": "{"parts":[{"ETag":"etag-blah","PartNumber":1}],"uploadId":"some_id","fileId":"mockSampleFileId","type":"sample"}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/features10x",
    {
      "body": "{"uploadStatus":"uploaded"}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "PATCH",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/barcodes10x",
    {
      "body": "{"uploadStatus":"uploaded"}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "PATCH",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/matrix10x",
    {
      "body": "{"uploadStatus":"uploaded"}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "PATCH",
    },
  ],
]
`;

exports[`processUpload Uploads and updates redux correctly when there are no errors with cellranger v3 1`] = `
[
  {
    "data": Blob {},
    "headers": {
      "Content-Type": "application/octet-stream",
    },
    "method": "put",
    "onUploadProgress": [Function],
    "signal": AbortSignal {},
    "url": "theSignedUrl",
  },
  {
    "data": Blob {},
    "headers": {
      "Content-Type": "application/octet-stream",
    },
    "method": "put",
    "onUploadProgress": [Function],
    "signal": AbortSignal {},
    "url": "theSignedUrl",
  },
  {
    "data": Blob {},
    "headers": {
      "Content-Type": "application/octet-stream",
    },
    "method": "put",
    "onUploadProgress": [Function],
    "signal": AbortSignal {},
    "url": "theSignedUrl",
  },
]
`;

exports[`processUpload Uploads and updates redux correctly when there are no errors with cellranger v3: fetch calls 1`] = `
[
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples",
    {
      "body": "[{"name":"WT13","sampleTechnology":"10x","options":{}}]",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/features10x",
    {
      "body": "{"sampleFileId":"mockSampleFileId","size":1024}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/barcodes10x",
    {
      "body": "{"sampleFileId":"mockSampleFileId","size":1024}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/matrix10x",
    {
      "body": "{"sampleFileId":"mockSampleFileId","size":1024}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/sampleFiles/mockSampleFileId/beginUpload",
    {
      "body": "{"size":1024,"metadata":{"cellranger_version":"v3"}}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/sampleFiles/mockSampleFileId/beginUpload",
    {
      "body": "{"size":1024,"metadata":{}}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/sampleFiles/mockSampleFileId/beginUpload",
    {
      "body": "{"size":1024,"metadata":{}}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/completeMultipartUpload",
    {
      "body": "{"parts":[{"ETag":"etag-blah","PartNumber":1}],"uploadId":"some_id","fileId":"mockSampleFileId","type":"sample"}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/completeMultipartUpload",
    {
      "body": "{"parts":[{"ETag":"etag-blah","PartNumber":1}],"uploadId":"some_id","fileId":"mockSampleFileId","type":"sample"}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/completeMultipartUpload",
    {
      "body": "{"parts":[{"ETag":"etag-blah","PartNumber":1}],"uploadId":"some_id","fileId":"mockSampleFileId","type":"sample"}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "POST",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/features10x",
    {
      "body": "{"uploadStatus":"uploaded"}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "PATCH",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/barcodes10x",
    {
      "body": "{"uploadStatus":"uploaded"}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "PATCH",
    },
  ],
  [
    "http://localhost:3000/v2/experiments/project-uuid/samples/mockSampleId/sampleFiles/matrix10x",
    {
      "body": "{"uploadStatus":"uploaded"}",
      "headers": {
        "Content-Type": "application/json",
      },
      "method": "PATCH",
    },
  ],
]
`;
